flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/JdbcXaSinkFunction.java

```diff
@@ -0,0 +1,367 @@
+        this(
+                new JdbcBatchingOutputFormat<>(
+                        xaFacade,
+                        executionOptions,
+                        context -> {
+                            Preconditions.checkState(
+                                    !context.getExecutionConfig().isObjectReuseEnabled(),
+                                    "objects can not be reused with JDBC sink function");
+                            return JdbcBatchStatementExecutor.simple(
+                                    sql, statementBuilder, Function.identity());
+                        },
+                        JdbcBatchingOutputFormat.RecordExtractor.identity()),
+                xaFacade,
+                XidGenerator.semanticXidGenerator(),
+                new XaSinkStateHandlerImpl(),
+                options,
+                new XaGroupOpsImpl(xaFacade));
+    }
+
+    /**
+     * Creates a {@link JdbcXaSinkFunction}.
+     *
+     * <p>All parameters must be {@link java.io.Serializable serializable}.
+     *
+     * @param format {@link JdbcBatchingOutputFormat} to write records with
+     * @param xaFacade {@link XaFacade} to manage XA transactions
+     * @param xidGenerator {@link XidGenerator} to generate new transaction ids
+     */
+    public JdbcXaSinkFunction(
+            JdbcBatchingOutputFormat<T, T, JdbcBatchStatementExecutor<T>> format,
+            XaFacade xaFacade,
+            XidGenerator xidGenerator,
+            XaSinkStateHandler stateHandler,
+            JdbcExactlyOnceOptions options,
+            XaGroupOps xaGroupOps) {
+        this.xaFacade = Preconditions.checkNotNull(xaFacade);
+        this.xidGenerator = Preconditions.checkNotNull(xidGenerator);
+        this.format = Preconditions.checkNotNull(format);
+        this.stateHandler = Preconditions.checkNotNull(stateHandler);
+        this.options = Preconditions.checkNotNull(options);
+        this.xaGroupOps = xaGroupOps;
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext context) throws Exception {
+        JdbcXaSinkFunctionState state = stateHandler.load(context);
+        hangingXids = new LinkedList<>(state.getHanging());
+        preparedXids = new ArrayList<>(state.getPrepared());
+        LOG.info(
+                "initialized state: prepared xids: {}, hanging xids: {}",
+                preparedXids.size(),
+                hangingXids.size());
+    }
+
+    @Override
+    public void open(Configuration configuration) throws Exception {
+        super.open(configuration);
+        xidGenerator.open();
+        xaFacade.open();
+        format.setRuntimeContext(getRuntimeContext());
+        format.open(
+                getRuntimeContext().getIndexOfThisSubtask(),
+                getRuntimeContext().getNumberOfParallelSubtasks());
+        hangingXids = new LinkedList<>(xaGroupOps.failOrRollback(hangingXids).getForRetry());
+        commitUpToCheckpoint(Optional.empty());
```

Comment:
I have some questions:
1. Will this leave some retry Xids in the `preparedXids`? 
2. If the job is failed, will the job restart the checkpoint from checkpoint: 0?
3. If yes, will the retry Xids in the `preparedXids` not be comitted for a long time?

Reply:
1. Yes, failed Xds will be returned to `preparedXids` again
2. The job will restart from the latest completed checkpoint if any (decided by `CheckpointCoordinator` )
3. Upon restart, `preparedXids` will be committed again (or erlier, if a checkpoint completion notification is received before restart)